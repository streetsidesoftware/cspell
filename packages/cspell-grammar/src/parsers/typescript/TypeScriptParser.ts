import { pipe, opFlatten, opMap } from '@cspell/cspell-pipe/sync';
import type { ParsedText, ParseResult, Scope, ScopeChain } from '@cspell/cspell-types/Parser';
import { compileGrammar } from '../..';
import { grammar } from '../../grammars/typescript';
import { appendMappedText } from '../../mappers/appendMappedText';
import { mapRawString } from '../../mappers/typescript';
import { createParser } from '../../parser/parser';
import { ScopePool } from '../../parser/scope';
import type { TokenizedLine } from '../../parser/types';

const tsGrammar = compileGrammar(grammar);

const pool = new ScopePool();

const useScope = new WeakMap<ScopeChain, boolean>();

function* transform(texts: ParseResult['parsedTexts']): ParseResult['parsedTexts'] {
    for (const parsed of texts) {
        if (doesScopeMatch(parsed.scope, 'constant.character.escape.ts')) {
            const mapped = mapRawString(parsed.text);
            const scope = parsed.scope ? pool.parseScope(parsed.scope) : undefined;
            yield {
                text: mapped.text,
                scope: scope?.parent,
                map: mapped.map,
                range: parsed.range,
            };
            continue;
        }
        yield parsed;
    }
}

function* mergeStringResults(results: Iterable<ParsedText>): Iterable<ParsedText> {
    let last: ParsedText | undefined;
    for (const next of results) {
        if (!doesScopeMatch(next.scope, 'string.')) {
            if (last) {
                yield last;
                last = undefined;
            }
            yield next;
            continue;
        }
        if (!last) {
            last = next;
            continue;
        }
        if (next.scope !== last.scope || last.range[1] !== next.range[0]) {
            yield last;
            last = next;
            continue;
        }
        last = mergeParsedText(last, next);
    }
    if (last) yield last;
}

function mergeParsedText(a: ParsedText, b: ParsedText): ParsedText {
    const abT = appendMappedText(a, b);
    const ab: ParsedText = {
        text: abT.text,
        scope: a.scope,
        range: [a.range[0], b.range[1]],
        map: abT.map,
        delegate: a.delegate,
    };

    return ab;
}

function filterScope(scope: ScopeChain): boolean {
    const cached = useScope.get(scope);
    if (cached !== undefined) return cached;
    const value = scope.value;
    const use = !value.startsWith('punctuation') && !value.startsWith('keyword.');
    useScope.set(scope, use);
    return use;
}

function mapTokenizedLine(tl: TokenizedLine): ParseResult['parsedTexts'] {
    return tl.tokens
        .filter((t) => filterScope(t.scope))
        .map((t) => ({
            text: t.text,
            range: [tl.offset + t.range[0], tl.offset + t.range[1]] as const,
            scope: t.scope,
        }));
}

function mapTokenizedLines(itl: Iterable<TokenizedLine>): ParseResult['parsedTexts'] {
    return pipe(itl, opMap(mapTokenizedLine), opFlatten(), transform, mergeStringResults);
}

export const parser = createParser(tsGrammar, 'typescript', mapTokenizedLines);

function doesScopeMatch(s: Scope | undefined, match: string): boolean {
    if (!s) return false;
    return typeof s === 'string' ? s.startsWith(match) : s.value.startsWith(match);
}
