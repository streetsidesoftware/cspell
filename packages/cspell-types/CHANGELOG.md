# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [5.5.2](https://github.com/streetsidesoftware/cspell/compare/v5.5.1...v5.5.2) (2021-05-30)


### Bug Fixes

* Ignore hash links like `[abcdefa]` ([#1293](https://github.com/streetsidesoftware/cspell/issues/1293)) ([888e25d](https://github.com/streetsidesoftware/cspell/commit/888e25d48254cc57c9fb4f3c5a65d57dfffacff0))





## [5.5.1](https://github.com/streetsidesoftware/cspell/compare/v5.5.0...v5.5.1) (2021-05-29)

**Note:** Version bump only for package @cspell/cspell-types





# [5.5.0](https://github.com/streetsidesoftware/cspell/compare/v5.4.1...v5.5.0) (2021-05-29)


### Features

* Remove incorrect Ignore Hex Digits Regexp ([#1277](https://github.com/streetsidesoftware/cspell/issues/1277)) ([2621eb0](https://github.com/streetsidesoftware/cspell/commit/2621eb02f487d9e466b4936bde8650c338b320b8)), closes [#1276](https://github.com/streetsidesoftware/cspell/issues/1276) [#1276](https://github.com/streetsidesoftware/cspell/issues/1276)





## [5.4.1](https://github.com/streetsidesoftware/cspell/compare/v5.4.0...v5.4.1) (2021-05-11)

**Note:** Version bump only for package @cspell/cspell-types





# [5.4.0](https://github.com/streetsidesoftware/cspell/compare/v5.3.12...v5.4.0) (2021-05-05)

**Note:** Version bump only for package @cspell/cspell-types





## [5.3.12](https://github.com/streetsidesoftware/cspell/compare/v5.3.11...v5.3.12) (2021-04-06)

**Note:** Version bump only for package @cspell/cspell-types





## [5.3.11](https://github.com/streetsidesoftware/cspell/compare/v5.3.10...v5.3.11) (2021-04-03)

**Note:** Version bump only for package @cspell/cspell-types





## [5.3.10](https://github.com/streetsidesoftware/cspell/compare/v5.3.9...v5.3.10) (2021-04-02)

**Note:** Version bump only for package @cspell/cspell-types





## [5.3.8](https://github.com/streetsidesoftware/cspell/compare/v5.3.7...v5.3.8) (2021-03-17)


### Bug Fixes

* update schema generator ([#1074](https://github.com/streetsidesoftware/cspell/issues/1074)) ([528b7d0](https://github.com/streetsidesoftware/cspell/commit/528b7d0b67e1f947832de24e7b66c3d7dfd5fb7f))





## [5.3.7](https://github.com/streetsidesoftware/cspell/compare/v5.3.7-alpha.3...v5.3.7) (2021-03-05)

**Note:** Version bump only for package @cspell/cspell-types





## [5.3.7-alpha.3](https://github.com/streetsidesoftware/cspell/compare/v5.3.7-alpha.2...v5.3.7-alpha.3) (2021-03-05)

**Note:** Version bump only for package @cspell/cspell-types





## [5.3.5](https://github.com/streetsidesoftware/cspell/compare/v5.3.4...v5.3.5) (2021-03-05)


### Bug Fixes

* fix schema pattern for language IDs. ([#1034](https://github.com/streetsidesoftware/cspell/issues/1034)) ([591dde2](https://github.com/streetsidesoftware/cspell/commit/591dde2532a7698f41c5bf824ead226a7ae78cfe))





## [5.3.4](https://github.com/streetsidesoftware/cspell/compare/v5.3.3...v5.3.4) (2021-03-01)

**Note:** Version bump only for package @cspell/cspell-types





## [5.3.3](https://github.com/streetsidesoftware/cspell/compare/v5.3.2...v5.3.3) (2021-02-26)

**Note:** Version bump only for package @cspell/cspell-types





## [5.3.2](https://github.com/streetsidesoftware/cspell/compare/v5.3.1...v5.3.2) (2021-02-26)

**Note:** Version bump only for package @cspell/cspell-types





## [5.3.1](https://github.com/streetsidesoftware/cspell/compare/v5.3.0...v5.3.1) (2021-02-25)

**Note:** Version bump only for package @cspell/cspell-types





# [5.3.0](https://github.com/streetsidesoftware/cspell/compare/v5.3.0-alpha.4...v5.3.0) (2021-02-25)

**Note:** Version bump only for package @cspell/cspell-types





# [5.3.0-alpha.4](https://github.com/streetsidesoftware/cspell/compare/v5.3.0-alpha.3...v5.3.0-alpha.4) (2021-02-25)


### Bug Fixes

* be able to report on glob source. ([#1001](https://github.com/streetsidesoftware/cspell/issues/1001)) ([1020d56](https://github.com/streetsidesoftware/cspell/commit/1020d56ba04e8418d162d4bde8e4948546a8f7d2))





# [5.3.0-alpha.3](https://github.com/streetsidesoftware/cspell/compare/v5.3.0-alpha.2...v5.3.0-alpha.3) (2021-02-23)


### Bug Fixes

* Improve reporting on files matching glob patterns. ([#994](https://github.com/streetsidesoftware/cspell/issues/994)) ([da991f9](https://github.com/streetsidesoftware/cspell/commit/da991f93a061c5b64ce437332c7107ef2ef89472))





# [5.3.0-alpha.2](https://github.com/streetsidesoftware/cspell/compare/v5.3.0-alpha.1...v5.3.0-alpha.2) (2021-02-22)


### Bug Fixes

* Add schema support for `enableFiletypes`. ([#979](https://github.com/streetsidesoftware/cspell/issues/979)) ([d18706e](https://github.com/streetsidesoftware/cspell/commit/d18706e3ffd7a6446396d339b85a753b4ae451e0))
* Add support for custom dictionaries ([#982](https://github.com/streetsidesoftware/cspell/issues/982)) ([196921d](https://github.com/streetsidesoftware/cspell/commit/196921dc635dce7cfdd5b6c300af54beeebde91a))





# [5.3.0-alpha.1](https://github.com/streetsidesoftware/cspell/compare/v5.3.0-alpha.0...v5.3.0-alpha.1) (2021-02-19)


### Bug Fixes

* Fix schema generation to use `deprecatedMessage` ([#972](https://github.com/streetsidesoftware/cspell/issues/972)) ([492dca9](https://github.com/streetsidesoftware/cspell/commit/492dca91466773bdf247fdb87f93d64914d5e3e1))





# [5.3.0-alpha.0](https://github.com/streetsidesoftware/cspell/compare/v5.2.4...v5.3.0-alpha.0) (2021-02-18)


### Features

* Be able to specify files to spell check within the config. ([#948](https://github.com/streetsidesoftware/cspell/issues/948)) ([23f7a48](https://github.com/streetsidesoftware/cspell/commit/23f7a488ef500fb1df5cd234c7d3c2ab4ec02961)), closes [#571](https://github.com/streetsidesoftware/cspell/issues/571)
* Part 1 - glob patterns are relative to the config file. ([#921](https://github.com/streetsidesoftware/cspell/issues/921)) ([a250448](https://github.com/streetsidesoftware/cspell/commit/a2504484ec38f15804cc0a203317266f83566b7c))
* Support local configuration files ([#966](https://github.com/streetsidesoftware/cspell/issues/966)) ([0ccc5fe](https://github.com/streetsidesoftware/cspell/commit/0ccc5fe9eb70ca3a4c6e5a3fc0b653465e76983c))





## [5.2.4](https://github.com/streetsidesoftware/cspell/compare/v5.2.3...v5.2.4) (2021-01-28)

**Note:** Version bump only for package @cspell/cspell-types

<!---
cspell: ignore abcdefa
--->
