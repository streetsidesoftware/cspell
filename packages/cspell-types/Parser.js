module.exports = require('./dist/Parser');
