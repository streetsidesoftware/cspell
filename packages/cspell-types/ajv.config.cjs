'use strict';

const keywords = ['deprecationMessage', 'markdownDescription', 'scope'];

function addKeywords(ajv) {
    for (const keyword of keywords) {
        ajv.addKeyword(keyword);
    }
}

module.exports = addKeywords;
