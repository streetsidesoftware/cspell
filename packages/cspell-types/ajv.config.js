'use strict';

const keywords = ['deprecatedMessage', 'markdownDescription', 'scope'];

function addKeywords(ajv) {
    for (const keyword of keywords) {
        ajv.addKeyword(keyword);
    }
}

module.exports = addKeywords;
