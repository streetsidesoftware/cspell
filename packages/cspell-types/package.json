{
  "name": "@cspell/cspell-types",
  "publishConfig": {
    "access": "public"
  },
  "version": "6.6.0",
  "description": "Types for cspell and cspell-lib",
  "type": "commonjs",
  "main": "dist/index.js",
  "typings": "dist/index.d.ts",
  "typesVersions": {
    "*": {
      "*": [
        "dist/*"
      ],
      "dist/index.d.ts": [
        "dist/index.d.ts"
      ],
      "index.d.ts": [
        "dist/index.d.ts"
      ]
    }
  },
  "exports": {
    ".": "./dist/index.js",
    "./Parser": "./dist/Parser.js",
    "./Parser.js": "./dist/Parser.js",
    "./cspell.schema.json": "./cspell.schema.json",
    "./ajv.config": "./ajv.config.js",
    "./ajv.config.js": "./ajv.config.js"
  },
  "scripts": {
    "clean": "rimraf dist coverage .tsbuildinfo",
    "build": "pnpm run compile && pnpm run build-schema",
    "build-dev": "tsc -p tsconfig.dev.json",
    "build-schema": "ts-json-schema-generator --no-top-ref --path src/CSpellSettingsDef.ts --type CSpellSettings --validation-keywords markdownDescription  --validation-keywords scope --validation-keywords deprecated --validation-keywords deprecationMessage -o  ./cspell.schema.json && cp ./cspell.schema.json ../..",
    "clean-build": "pnpm run clean && pnpm run build",
    "compile": "tsc -p .",
    "test-schema": "ajv -s ./cspell.schema.json -d \"cspell.test.{json,yaml}\" -c ./ajv.config.js",
    "test": "jest && pnpm run test-schema",
    "prepare": "echo Skip Building Types"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/streetsidesoftware/cspell.git"
  },
  "files": [
    "dist",
    "ajv.config.js",
    "cspell.schema.json",
    "!**/__mocks__",
    "!**/*.test.*",
    "!**/*.spec.*",
    "!**/*.map"
  ],
  "keywords": [
    "cspell",
    "types"
  ],
  "author": "Jason Dent",
  "license": "MIT",
  "engines": {
    "node": ">=14"
  },
  "devDependencies": {
    "ajv-cli": "^5.0.0",
    "jest": "^28.1.3",
    "rimraf": "^3.0.2",
    "ts-json-schema-generator": "^1.0.0",
    "typescript": "^4.7.4"
  }
}
