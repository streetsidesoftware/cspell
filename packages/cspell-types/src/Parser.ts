export type ParserOptions = Record<string, unknown>;

export type ParserName = string;

export interface Parser {
    /** Name of parser */
    name: ParserName;
    /**
     * Parse Method
     * @param content - full content of the file
     * @param filename - filename
     */
    parse(content: string, filename: string): ParseResult;
}

export interface ParseResult {
    content: string;
    filename: string;
    parsedTexts: Iterable<ParsedText>;
}

export interface ParsedText {
    /**
     * The text extracted and possibly transformed
     */
    text: string;
    /**
     * start and end offsets of the text
     */
    range: Range;
    /**
     * The Scope annotation for a segment of text.
     * Used by the spell checker to apply spell checking options
     * based upon the value of the scope.
     */
    scope?: Scope | undefined;
    /**
     * The source map is used to support text transformations.
     *
     * See: {@link SourceMap}
     */
    map?: SourceMap | undefined;
    /**
     * Used to delegate parsing the contents of `text` to another parser.
     *
     */
    delegate?: DelegateInfo | undefined;
}

/**
 * A SourceMap is used to map transform a piece of text back to its original text.
 * This is necessary in order to report the correct location of a spelling issue.
 * An empty source map indicates that it was a 1:1 transformation.
 *
 * The values in a source map are number pairs (even, odd) relative to the beginning of each
 * string segment.
 * - even - offset in the source text
 * - odd - offset in the transformed text
 *
 * Offsets start a 0
 *
 * Example:
 *
 * - Original text: `Grand Caf\u00e9 Bj\u00f8rvika`
 * - Transformed text: `Grand Café Bjørvika`
 * - Map: [9, 9, 15, 10, 18, 13, 24, 14]
 *
 * | offset | original    | offset | transformed |
 * | ------ | ----------- | ------ | ----------- |
 * | 0-9    | `Grand Caf` | 0-9    | `Grand Caf` |
 * | 9-15   | `\u00e9`    | 9-10   | `é`         |
 * | 15-18  | ` Bj`       | 10-13  | ` Bj`       |
 * | 18-24  | `\u00f8`    | 13-14  | `ø`         |
 * | 24-29  | `rvika`     | 14-19  | `rvika`     |
 *
 * <!--- cspell:ignore Bjørvika rvika --->
 */
export type SourceMap = number[];

export type Range = readonly [start: number, end: number];

/**
 * DelegateInfo is used by a parser to delegate parsing a subsection of a document to
 * another parser. The following information is used by the spell checker to match
 * the parser.
 */
export interface DelegateInfo {
    /**
     * Proposed virtual file name including the extension.
     * Format: `./${source_filename}/${block_number}.${ext}
     * Example: `./README.md/1.js`
     */
    filename: string;
    /**
     * The filename of the origin of the virtual file block.
     * Example: `./README.md`
     */
    originFilename: string;
    /**
     * Proposed file extension
     * Example: `.js`
     */
    extension: string;
    /**
     * Filetype to use
     * Example: `javascript`
     */
    fileType: string;
}

/**
 * Scope - chain of scope going from local to global
 *
 * Example:
 * ```
 * `comment.block.documentation.ts` -> `meta.interface.ts` -> `source.ts`
 * ```
 */
export interface ScopeChain {
    readonly value: string;
    readonly parent?: ScopeChain | undefined;
}

/**
 * A string representing a scope chain separated by spaces
 *
 * Example: `comment.block.documentation.ts meta.interface.ts source.ts`
 */
export type ScopeString = string;

export type Scope = ScopeChain | ScopeString;
