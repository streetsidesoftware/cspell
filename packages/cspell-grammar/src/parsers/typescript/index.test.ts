import { parser } from './TypeScriptParser';
import { promises as fs } from 'fs';
import * as path from 'path';
import { ParseResult } from '@cspell/cspell-types/Parser';

const fixtures = path.join(__dirname, '../../../fixtures');

describe('TypeScript Parser', () => {
    const sample1File = 'TypeScript/escape-codes.ts';
    test('parser', () => {
        expect(parser.name).toBe('typescript');
    });

    test('parse sample', async () => {
        const content = await readSample(sample1File);
        const p = parser.parse(content, sample1File);
        expect(stringifyResult(p)).toMatchSnapshot();
    });
});

function readSample(filename: string): Promise<string> {
    return fs.readFile(path.resolve(fixtures, filename), 'utf-8');
}

function stringifyResult(result: ParseResult): string {
    function mapParsedTexts(t: ParseResult['parsedTexts']): string[] {
        return [...t].map((p) => `${p.range[0]}-${p.range[1]} ${JSON.stringify(p.text)} ${p.scope?.toString() ?? ''}`);
    }

    return `\
filename: ${result.filename}
parsed:
${mapParsedTexts(result.parsedTexts).join('\n')}
`;
}
