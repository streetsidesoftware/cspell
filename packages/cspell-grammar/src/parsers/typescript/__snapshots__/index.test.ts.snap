// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TypeScript Parser parse TypeScript/escape-codes.ts 1`] = `
"filename: TypeScript/escape-codes.ts
parsed:
6-12 \\"places\\" identifier code.ts source.ts
17-26 \\"Amsterdam\\" string.quoted.single.ts code.ts code.ts meta.brace.ts code.ts source.ts
30-40 \\"Orléans\\" string.quoted.single.ts code.ts code.ts meta.brace.ts code.ts source.ts
51-57 \\"sample\\" identifier code.ts source.ts
66-70 \\"text\\" identifier code.ts code.ts meta.brace.ts code.ts source.ts
73-162 \\"Fingerspitzengefühl is a German term.\\\\nIt’s pronounced as follows: [ˈfɪŋɐˌʃpɪtsənɡəˌfyːl]\\" string.quoted.single.ts code.ts code.ts meta.brace.ts code.ts source.ts
169-172 \\"hex\\" identifier code.ts code.ts meta.brace.ts code.ts source.ts
175-557 \\"Fingerspitzengefühl is a German term.\\\\nIt’s pronounced as follows: [ˈfɪŋɐˌʃpɪtsənɡəˌfyːl]\\" string.quoted.single.ts code.ts code.ts meta.brace.ts code.ts source.ts
564-569 \\"mixed\\" identifier code.ts code.ts meta.brace.ts code.ts source.ts
572-731 \\"Fingerspitzengefühl is a German term.\\\\nIt’s pronounced as follows: [ˈfɪŋɐˌʃpɪtsənɡəˌfyːl]\\" string.quoted.single.ts code.ts code.ts meta.brace.ts code.ts source.ts
745-749 \\".log\\" identifier code.ts source.ts
750-756 \\"places\\" identifier code.ts code.ts meta.brace.ts code.ts source.ts
766-770 \\".log\\" identifier code.ts source.ts
772-774 \\"%o\\" string.quoted.single.ts code.ts code.ts meta.brace.ts code.ts source.ts
777-783 \\"sample\\" identifier code.ts code.ts meta.brace.ts code.ts source.ts
"
`;

exports[`TypeScript Parser parse TypeScript/sample1.ts 1`] = `
"filename: TypeScript/sample1.ts
parsed:
3-4 \\"\\\\n\\" comment.block.documentation.ts code.ts source.ts
4-32 \\" * This is some sample code\\\\n\\" comment.block.documentation.ts code.ts source.ts
32-33 \\" \\" comment.block.documentation.ts code.ts source.ts
50-60 \\"regExpPath\\" identifier code.ts source.ts
64-65 \\"(\\" regexp.ts code.ts source.ts
65-67 \\"\\\\\\\\w\\" escape.regexp.ts regexp.ts code.ts source.ts
67-68 \\"+\\" regexp.ts code.ts source.ts
68-70 \\"\\\\\\\\/\\" escape.regexp.ts regexp.ts code.ts source.ts
70-72 \\")+\\" regexp.ts code.ts source.ts
89-100 \\"regExpPath2\\" identifier code.ts source.ts
104-105 \\"(\\" regexp.ts code.ts source.ts
105-107 \\"\\\\\\\\w\\" escape.regexp.ts regexp.ts code.ts source.ts
107-108 \\"+\\" regexp.ts code.ts source.ts
108-109 \\"[\\" brace.regexp.ts regexp.ts code.ts source.ts
109-110 \\"/\\" character-class.regexp.ts brace.regexp.ts regexp.ts code.ts source.ts
110-112 \\"\\\\\\\\\\\\\\\\\\" escape.regexp.ts character-class.regexp.ts brace.regexp.ts regexp.ts code.ts source.ts
112-113 \\"]\\" brace.regexp.ts regexp.ts code.ts source.ts
113-115 \\")+\\" regexp.ts code.ts source.ts
133-140 \\"comment\\" identifier code.ts source.ts
144-168 \\"This is a comment string\\" string.quoted.single.ts code.ts source.ts
"
`;
